### PR TITLE
fix: consistent rule doc notices and sections

### PR DESCRIPTION
**Pull request by bmish**
This PR adds the following standardized rule doc notices:

* 💼 This rule is enabled in the following [configs](https://github.com/jest-community/eslint-plugin-jest#shareable-configurations): `all`, `recommended`.
* 🔧 This rule is automatically fixable using the `--fix` [flag](https://eslint.org/docs/latest/user-guide/command-line-interface#--fix) on the command line.
* 💡 This rule provides [suggestions](https://eslint.org/docs/developer-guide/working-with-rules#providing-suggestions) that can be applied manually.
* ❌ This rule is deprecated.

This improves usability by enabling users to clearly and easily find out if a rule is applying to their codebase given their configuration as well as other key rule details like the availability of auto-fixers or auto-suggestions that can help with addressing violations. 

All these notices are tested to ensure they are in the correct position and not forgotten. `yarn tools:regenerate-docs` will also automatically add or update these notices in each rule doc (some inspiration taken from eslint-plugin-unicorn which has a similar command). 

It's common for ESLint and many popular ESLint plugins to include notices like these. Some examples (note: I am a maintainer on some of these plugins and have been working to standardize rule documentation):
* https://eslint.org/docs/latest/rules/no-unused-vars
* https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/no-unknown-property.md
* https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/consistent-destructuring.md
* https://github.com/ember-cli/eslint-plugin-ember/blob/master/docs/rules/no-get.md
* https://github.com/not-an-aardvark/eslint-plugin-eslint-plugin/blob/main/docs/rules/require-meta-schema.md

I also added tests about a few other aspects of rule docs:
* The title matches the rule description
* Has a `Rule details` section
* Has an `Options` section only if the rule has options and mentions any named options

### CODE REVIEW

**Review by G-Rath — CHANGES_REQUESTED** [review]
I'll give this a more detailed review later but we've already got a tool for updating our docs which runs in CI so it already covers at least some of what you've added tests for, and automatically does the fixing so want to favor that - would you be interested in expanding that to generate what you've manually added to these docs?

I love the idea of standardizing rule documentation but am wondering if it would be best to try and submit some sort of loose RFC/template so that we could see all the different parts + make it easy to write a tool to help generate them.

@MichaelDeBoey this feels like the sort of thing that would tie in fantastically with the new ESLint community stuff that got landed, so interested in if you have thoughts.

**Comment by bmish** [comment]
@G-Rath thanks for your initial thoughts.

1. Yes, I can update the rule doc generator to generate these notices as well. That will take some refactoring to make the generator a bit more sophisticated, so I'll work on that over the next week. If you like the updates to the rule docs, it could be worth be merging this first to avoid potential merge conflicts, and then I could create a separate dedicated follow-up PR to focus only on switching to automatic notice generation.
2. I'm definitely a fan of standardizing the rule documentation as I called out in the PR description. But I'm not sure an ESLint RFC would be appropriate as the ESLint team isn't really in the business of enforcing a documentation standard for plugins. However, we do have [generator-eslint](https://github.com/eslint/generator-eslint), and I like the idea of updating its [rule doc template](https://github.com/eslint/generator-eslint/blob/main/rule/templates/_doc.md) to include notices, so I will plan to do that.

File: src/__tests__/rules.test.ts
Comment (SimenB) [inline src/__tests__/rules.test.ts:123]: should use `it.each`: https://jestjs.io/docs/api#testeachtablename-fn-timeout

Comment (bmish) [reply src/__tests__/rules.test.ts]: Neat, updated to use that.

**Comment by bmish** [comment]
@G-Rath good news, I went ahead and updated the `yarn tools:regenerate-docs` command to automatically add or update the rule notices in each rule doc.

File: README.md
Comment (G-Rath) [inline README.md:218]: @SimenB I guess we forgot to remove this in the new major 😅

File: docs/rules/consistent-test-it.md
Comment (G-Rath) [inline docs/rules/consistent-test-it.md:5]: It should be safe to apply prettier to these and I think it's better to be consistent as much as possible

File: docs/rules/consistent-test-it.md
Comment (G-Rath) [inline docs/rules/consistent-test-it.md:11]: I'd really love it if we could somehow not need these markers all together, but I can't think of a good way to match that - still, since we're wanting these to always be at top and everything until this line is generated I think we can just use a "end rule notices" marker (I'd like that name too, because its consistent with the format of the markers in the readme)

I think it would also make sense to have the tool add this if it can't find the marker, which would make it a bit smoother if you're making a new rule.

File: docs/rules/expect-expect.md
Comment (G-Rath) [inline docs/rules/expect-expect.md:12]: I think we should actually be ensuring `details` is lowercase because that's more consistent with other headers (and technically more grammatically correct)

File: docs/rules/no-done-callback.md
Comment (G-Rath) [inline docs/rules/no-done-callback.md:9]: nit: what do you think of "This rule provides suggestions that can be applied manually"? That feels a bit more grammatically natural to me

File: docs/rules/no-if.md
Comment (G-Rath) [inline docs/rules/no-if.md:5]: nit: doesn't have to be done in this PR, but a good followup could be adding "in favor of ..." if there's a suggested replacement

File: README.md
Comment (bmish) [reply README.md:218]: Yeah, luckily with this change it will be clearly visible that it still exists but is deprecated.

File: src/__tests__/rules.test.ts
Comment (G-Rath) [inline src/__tests__/rules.test.ts:87]: I don't really think we need this test because we're able to re-run the docs tool in ci and check if any files have changed, whereas we'd have to update this test if the output changes and its relatively complex.

The only thing we won't catch from the tool is if we've got or not got a header for options, but I'm not super fussed about that because that's a pretty obvious one and the more important thing is ensuring all the options are documented which is a lot harder to automate. (it'd also be pretty easy to have the docs tool check & fail on anyway).

File: src/__tests__/rules.test.ts
Comment (G-Rath) [inline src/__tests__/rules.test.ts:130]: it doesn't matter since we'll be axing this test, but fwiw you should be able to use `not.toContain` here

File: tools/regenerate-docs.ts
Comment (G-Rath) [inline tools/regenerate-docs.ts:166]: Relating to my comment above since we're always working with the very start of the file, I think we should replace this with `removeHeader` + `generateHeader` functions, and then the `generateHeader` function can be simplified by just being an array:

```
const generateHeader = () => {
  const header = [`# ${description} (\`${name}\`)`];

  if (/* rule is in a config */) {
    header.push('', '$ this rule is in configs ...');
  }

  // etc

  return header.join('\n');
}
```

**Review by G-Rath — CHANGES_REQUESTED** [review]
Looks like a good start - I've put some comments around on how I'd like the output to change; I'll hold off reviewing the `tools/` stuff in detail until we've gotten the actual output sorted because that'll impact the actual implementation :)

File: README.md
Comment (G-Rath) [reply README.md:218]: I guess - alternatively you could say now its clearly visible that it exists at all even though we don't want people to be using it 🤷

**Comment by G-Rath** [comment]
> I'm not sure an ESLint RFC would be appropriate as the ESLint team isn't really in the business of enforcing a documentation standard for plugins

No but they have rules themselves that have documentation, and this is also exactly why @eslint-community has been created. I don't think it needs to be a strictly large or full-blown RFC but I do like the idea of something like this going through the RFC process in terms of visibility and discussion as I think it could help unify the community and especially help people that are looking to create their own plugins (I think especially for non-native English speakers, having a template like that would make it a lot easier).

cc @nzakas cause you're probably the best person to give a steer.

**Comment by SimenB** [comment]
This is awesome @bmish, thanks for taking the time! Is the plan to upstream this into some shared eslint tool?

File: README.md
Comment (bmish) [reply README.md:218]: My personal preference is to continue documenting deprecated features while clearly indicating they are deprecated. People may still be using the deprecated features and need to find their documentation, and omitting any link to them can cause confusion.

File: docs/rules/no-done-callback.md
Comment (bmish) [reply docs/rules/no-done-callback.md]: That's fine, updated.

File: docs/rules/expect-expect.md
Comment (bmish) [reply docs/rules/expect-expect.md]: Sounds good, updated.

File: docs/rules/consistent-test-it.md
Comment (bmish) [reply docs/rules/consistent-test-it.md]: I was able to remove the prettier ignore comments.

File: docs/rules/consistent-test-it.md
Comment (bmish) [reply docs/rules/consistent-test-it.md]: I switched to using an "end rule header" marker only. And the tool will update the header or create it if it doesn't exist.

File: src/__tests__/rules.test.ts
Comment (bmish) [reply src/__tests__/rules.test.ts:82]: I removed the tests for the content that is automatically generated.

I do want to continue to advocate for testing that:
* incorrect notices are not present
* section title consistency like "Rule details"
* "Options" section is included only for rules with options

Not only did these tests catch many issues with the jest rule docs, but they also save a lot of time during PR reviews, since many contributors will often just copy and paste an existing rule doc, and then PR reviewers will have to remind them to remove all the outdated content (and outdated content/notices can easily slip through). These tests have saved me a lot of time in my other plugins.

I also have some tests in other plugins that ensure that each configuration option is mentioned in the rule doc like [this example](https://github.com/ember-cli/eslint-plugin-ember/blob/cb145d8f1b37cfa940f27ce57b0ffe45dc85ab4b/tests/rule-setup.js#L133). That would be a good follow-up to this PR, so I do think we can and should have at least basic tests that rule docs document their rule options.

I would also argue that having a few actual jest tests for rule docs is preferable to having the rule doc generator fail. Jest tests/assertions are standardized, easy/convenient to write, and will provide more detailed and consistent failure output than some custom failure in the rule doc generator.

File: tools/regenerate-docs.ts
Comment (bmish) [reply tools/regenerate-docs.ts]: I went ahead and refactored to something like this -- `generateRuleHeaderLines()` + `replaceOrCreateHeader()`

**Comment by bmish** [comment]
> > I'm not sure an ESLint RFC would be appropriate as the ESLint team isn't really in the business of enforcing a documentation standard for plugins
> 
> No but they have rules themselves that have documentation, and this is also exactly why @eslint-community has been created. I don't think it needs to be a strictly large or full-blown RFC but I do like the idea of something like this going through the RFC process in terms of visibility and discussion as I think it could help unify the community and especially help people that are looking to create their own plugins (I think especially for non-native English speakers, having a template like that would make it a lot easier).
 
> This is awesome @bmish, thanks for taking the time! Is the plan to upstream this into some shared eslint tool?

Here's a summary of my plans to standardize ESLint plugin documentation:

1. Submitting PRs like this one to ensure the top ESLint plugins conform to basic documentation standards (as mentioned in the PR description, I've already completed this for many of the top plugins). The documentation of ESLint and its top plugins will naturally serve as examples for other plugin authors to follow. 
2. Updating the documentation templates in [generator-eslint](https://github.com/eslint/generator-eslint) which can be used for generating a new plugin or rule.
3. Exploring creating a lint rule(s) in [eslint-plugin-eslint-plugin](https://github.com/not-an-aardvark/eslint-plugin-eslint-plugin) for enforcing documentation consistency (title, notices, etc), as most ESLint plugins already use this plugin and would thus easily pull in the new checks. We could also use a lint rule autofixer to generate the auto-generated parts of the rule doc (title, notices).
4. I'm open to working on some ESLint community recommendations around this, perhaps to point people to the generator-eslint templates or examples of high-quality / standardized documentation.

File: src/__tests__/rules.test.ts
Comment (G-Rath) [reply src/__tests__/rules.test.ts:82]: For now I disagree: most of what you've put is already covered [by us running the docs tool in CI](https://github.com/jest-community/eslint-plugin-jest/blob/main/.github/workflows/nodejs.yml#L103-L121), with the resulting output telling people what they need to do to resolve the problem (run the tool).

My dislike is with how brittle this still feels + you needing to export stuff from `tools/`, vs our current CI setup.

Comment (bmish) [reply src/__tests__/rules.test.ts:82]: I agree `tools/` isn't the best place to export stuff from so I can definitely move `rule-notices.ts` to `lib/` instead if you prefer.

But I'm confused by "most of what you've put is already covered [by us running the docs tool in CI](https://github.com/jest-community/eslint-plugin-jest/blob/main/.github/workflows/nodejs.yml#L103-L121)" as I've explicitly removed the tests for auto-generated content and only left in place tests that basic details are included or not included in the rule doc (which cannot be resolved by running the tool).

File: tools/regenerate-docs.ts
Comment (G-Rath) [inline tools/regenerate-docs.ts:116]: We might as well just return an array here:

```suggestion
const generateRuleHeaderLines = (description, name): string[] => [
  `# ${description} (\`${name}\`)`,
  ...getRuleNoticeLines(name),
  END_RULE_HEADER_MARKER,
];
```

File: src/__tests__/rules.test.ts
Comment (G-Rath) [reply src/__tests__/rules.test.ts:82]: I would expect the tool to be able to handle the notices completely, since if we can detect them we should be able to remove them as well.

As the other stuff it feels very low gain because the important part is actually _where_ in the document they are - i.e. you can make the tests pass by just adding those headers at the start or end. Personally what I'd rather see from this is actually further evolving the docs tool (e.g. have it output warnings about what you're testing for here) because that's more portable - we actually should already be almost at a point where you could have it as its own package (we'd just need to sort out loading the rules).

But that's out of scope for this I think - the key thing for me here is that we're still checking about the notices when we should be able to handle that in the tool.

File: tools/regenerate-docs.ts
Comment (G-Rath) [inline tools/regenerate-docs.ts:13]: I'm in favor of omitting the `generated by..` because I think its easy to locate the tool by searching for the header and the command might change (i.e. if we extracted it into a cli package, then what do we have here?)

File: tools/rule-notices.ts
Comment (G-Rath) [inline tools/rule-notices.ts:21]: I'd rather not have the `@ts-ignore` below instead of this as I don't think there's a lot of value vs using `string` & co; it might be possible to make it work but I'd have to play around with the types a bit (which I don't have the time for right now) so if you can't make it work lets just go with `string`.

File: tools/regenerate-docs.ts
Comment (G-Rath) [inline tools/regenerate-docs.ts:125]: It looks like maybe our typechecking is not running properly on these files - I'll look into that later, but you need to have explicit types on these params as currently they're implicitly `any`.

(You can make typechecking work by adding `tools/**/*` to the `include` array in `tsconfig.json`, but I need to check if that will impact the compiled output before actually committing that)

File: tools/rule-notices.ts
Comment (G-Rath) [inline tools/rule-notices.ts:92]: nit: I'd prefer using `for ... of`

File: tools/rule-notices.ts
Comment (G-Rath) [inline tools/rule-notices.ts:56]: I'm hoping that by moving the handling of notices entirely into the tool we can re-implement this to leverage native data structures a bit more rather than need all these elses i.e. we could have a map of `{ message: expected }`:

```
{
  [MESSAGE_TYPE.DEPRECATED]: rule.meta.deprecated,
}
```

which you could iterate over with something like:

```
for (const [notice, expected] of Object.entries(notices)) {
  if (expected) {
    header.push(notice);

    continue;
  }

  // or maybe linesToRemove.push(notice), and then do that in the outer function
  content.replace(notice, '');
}
```

(this is very psudo-y; let me know if that doesn't make sense 😅)

File: tools/regenerate-docs.ts
Comment (bmish) [reply tools/regenerate-docs.ts]: Good point, fixed.

File: tools/rule-notices.ts
Comment (bmish) [reply tools/rule-notices.ts]: Fixed. I recommend we enable eslint-plugin-unicorn recommended rules including [unicorn/no-array-for-each](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-array-for-each.md) to catch stuff like this.

File: tools/regenerate-docs.ts
Comment (bmish) [reply tools/regenerate-docs.ts]: Good catch, added types and updated tsconfig.json. We have `"noEmit": true` so I don't think this will cause any output to be produced?

File: tools/rule-notices.ts
Comment (G-Rath) [reply tools/rule-notices.ts]: I'm not committed to it enough yet to bring in a whole new plugin especially just for one rule (+ I don't think its worth the energy bikeshedding over all the rules we could have, given how stable this plugin is) - its more something I've been evaluating for the last couple of months to see how I feel about it :)

File: tools/regenerate-docs.ts
Comment (G-Rath) [reply tools/regenerate-docs.ts]: Thanks, but I'd really like to avoid changing the `tsconfig.json` in anyway in this PR just to be on the safe side - its also an indicator that I should do a look over the whole codebase for a bit of a tidy up

File: tools/rule-notices.ts
Comment (bmish) [reply tools/rule-notices.ts]: Well this is just one example. Of course I think there are many valuable rules in eslint-plugin-unicorn. Something to consider separately.

File: tools/regenerate-docs.ts
Comment (bmish) [reply tools/regenerate-docs.ts]: Okay removed change to `tsconfig.json`.

File: tools/rule-notices.ts
Comment (bmish) [reply tools/rule-notices.ts]: This `RuleName` type I have already evaluates to a `string`. I went ahead and removed this type since it's not really necessary, but this does not solve the `@ts-ignore` issue below.

```
Element implicitly has an 'any' type because expression of type '`jest/${string}`' can't be used to index type 'Record<string, RuleLevel> | { 'jest/no-alias-methods': string; 'jest/prefer-to-be': string; 'jest/prefer-to-contain': string; 'jest/prefer-to-have-length': string; }'.ts(7053)
```

This could be solved if we had a static type containing all the possible rule names, but I don't think we have that, and having to define a type containing all the rule names would be a maintenance burden, so it seems best to just use `@ts-ignore` for this situation.

File: tools/regenerate-docs.ts
Comment (bmish) [reply tools/regenerate-docs.ts]: Removed this part of the marker.

File: tools/rule-notices.ts
Comment (bmish) [reply tools/rule-notices.ts:55]: I switched over to having `getNoticesForRule()` return an object. I agree this is much nicer.

File: src/__tests__/rules.test.ts
Comment (bmish) [reply src/__tests__/rules.test.ts]: Okay, I went ahead and moved these tests (checking for `## Rule details` and `## Options`) into the generator. If these sections are incorrectly present, we will output the issue and the script will exit with failure.

File: docs/rules/no-if.md
Comment (bmish) [reply docs/rules/no-if.md]: Definitely like this idea so I went ahead and updated the notice to link to the replacement rule(s).

File: docs/rules/consistent-test-it.md
Comment (SimenB) [inline docs/rules/consistent-test-it.md:4]: can we link to the `README.md` file? Otherwise the link is broken on mobile

File: README.md
Comment (SimenB) [inline README.md:218]: nice, listing out all rules even if deprecated makes sense to me 👍

File: docs/rules/consistent-test-it.md
Comment (SimenB) [inline docs/rules/consistent-test-it.md:4]: also, should it say `config` instead of `configs` when only in a single one?

File: package.json
Comment (SimenB) [inline package.json:86]: what isn't covered?

Comment (G-Rath) [reply package.json:86]: It's the tools - becuase they're not imported by anything that is involved in the tests, they're not expected to be covered. When @bmish introduced tests that imported from there, they got included in coverage.

Now that those tests have been axed again, this can go too

File: docs/rules/consistent-test-it.md
Comment (bmish) [reply docs/rules/consistent-test-it.md]: Good point, updated.

File: package.json
Comment (bmish) [reply package.json]: Removed.

File: docs/rules/consistent-test-it.md
Comment (bmish) [reply docs/rules/consistent-test-it.md]: I could go either way on this. My first thought was that it was grammatically correct to use `configs` regardless, since this sentence is separated from the serial list with a colon, and it's clear that the list of configs is variable. If I'm wrong or you still prefer the other way, let me know.

Comment (SimenB) [reply docs/rules/consistent-test-it.md]: I'm not a native speaker, so happy to leave it 👍

**Review by SimenB — APPROVED** [review]

**Comment by G-Rath** [comment]
@bmish I've just landed #1239 which updates the `tsconfig.json` to check tools - could you rebase to include that, as `typecheck` is failing now.

File: tools/rule-notices.ts
Comment (G-Rath) [inline tools/rule-notices.ts:33]: I'm hoping to get this resolved via #1241 - in the meantime, disabling this rule is not the appropriate thing to do; instead you should use `@ts-expect-error` with a reason:

```suggestion
    // @ts-expect-error -- we don't have a static type for rule names
```

**Review by G-Rath — CHANGES_REQUESTED** [review]
This is looking awesome! If it's ok with you, I'm keen to toy with this a bit once its landed to see if it can be expanded into a self-standing tool.

We're just got a few type related things to resolve before landing this :)

File: tools/rule-notices.ts
Comment (SimenB) [reply tools/rule-notices.ts:33]: Merged in `main`, is this still right?

Comment (G-Rath) [reply tools/rule-notices.ts:33]: na now it can go completely

Comment (SimenB) [inline tools/rule-notices.ts:33]:
```suggestion
```

**Comment by SimenB** [comment]
Thanks @bmish, this is awesome!

**Comment by bmish** [comment]
@G-Rath @SimenB thanks a lot for helping to get this in! As part of my goal to standardize ESLint plugin documentation, I agree we should work to share this tool, and I'd be interested to collaborate on that. 

If we went with the CLI command here, it could possibly make sense for it to live inside [generator-eslint](https://github.com/eslint/generator-eslint), although that tool currently only has Yeoman generators in it.

I also plan to experiment with whether I could create a lint rule in [eslint-plugin-eslint-plugin](https://github.com/not-an-aardvark/eslint-plugin-eslint-plugin) for enforcing documentation consistency (title, notices, options, etc), as most ESLint plugins already use this plugin and would thus easily pull in the new checks. We could also use a lint rule autofixer to generate the auto-generated parts of the rule doc (title, notices).

File: docs/rules/consistent-test-it.md
Comment (ljharb) [reply docs/rules/consistent-test-it.md]: The link works fine for me on mobile, as-is.

**Pull request merged**
